### PR TITLE
Less constrained dependencies

### PR DESCRIPTION
**Pull request by micheelengronne**
Dependencies are currently too constrained and conflict with other ansible projects dependencies.

#### PR Type

- Bugfix Pull Request for https://github.com/ansible/molecule/issues/2023

### CODE REVIEW

**Comment by micheelengronne** [comment]
Agreed, except for pyyaml that goes from major version 3 to 5.

**Comment by micheelengronne** [comment]
I added an upper limit. But it is the widest as it takes only major version into account. Should it be reduced to minor as well ?

**Comment by decentral1se** [comment]
> But it is the widest as it takes only major version into account. 

As far as I know, that's what we've done so far. Thanks!

**Comment by micheelengronne** [comment]
You welcome. As it is currently blocking my stack and my roles development, do you know the next release date where this merge will be ?

**Comment by decentral1se** [comment]
Not sure. https://github.com/ansible/molecule/issues/2022 will block the CI, I think ...

**Comment by micheelengronne** [comment]
Hum, CI failed because of configuration issues in the travis file unrelated to that MR.

Weird.

**Comment by micheelengronne** [comment]
My bad. I misread the tests results. The error is still not related to this MR but is what https://github.com/ansible/molecule/pull/2025 tries to solve.

**Comment by webknjaz** [comment]
@micheelengronne as far as I remember, I kept those specific deps pinned because they don't follow semver.

Have you checked that every each of them is following semver? If not, you need to identify their breaking changes policies and set version boundaries accordingly.

**Comment by micheelengronne** [comment]
I understand your philosophy but `PyYAML` and `Jinja2` with these versions are required by ansible itself on ubuntu:bionic. To keep these 2 pined breaks the installation of molecule with ansible making this project useless (how can we test ansible roles without ansible ?).

`six` with this version is required by ansible-lint on ubuntu:bionic that molecule uses.

`tabulate` with this version is required by ansible-bender on ubuntu:bionic so we can have a separate environment for ansible-bender and molecule as they do not interact.

I am totally with you on ensuring that SemVer is respected but molecule is not the main project (ansible is) so these rules should be enforced at ansible level and for all its subprojects.

From my point of view, molecule should rather prioritizing on compatibility with ansible.

In the current state, molecule is perfectly useless on ubuntu:bionic as we can not have it installed along ansible.

The Semver rules I put in the setup correspond to the current state of ansible required dependencies. I fixed the upper limit to the current major version but if I understand correctly ansible did not so we might need to change the upper limit as well.

**Comment by micheelengronne** [comment]
As a side note, molecule does not work at all on debian:stretch but it is the Debian policy that keeps a completely outdated python3.

**Comment by decentral1se** [comment]
> Have you checked that every each of them is following semver?

Ah, good point. Following from https://github.com/ansible/molecule/issues/2022#issuecomment-488682442, we might upper bound Cerberus on minor versions. To revise my comment above, yes, we need to be sure that projects are following semver if we're doing this (my assumption ...).

**Comment by webknjaz** [comment]
@decentral1se exactly, when I was converting deps to be stored in `setup.cfg` I only set bounds based on the research of their policies.

We can still set upper bounds for non-semver compliant dists but we have to apply different rules (as in not using just a major version) for such limits.

@micheelengronne 
> but molecule is not the main project (ansible is) so these rules should be enforced at ansible level and for all its subprojects.

Nope. We are talking about the direct runtime deps of Molecule, not Ansible's. Ansible deps are declared there and those are transitive deps for Molecule unless specified otherwise.

**Comment by micheelengronne** [comment]
@webknjaz hum, you mean that if I do not declare ansible itself in my requirements.txt, molecule will install a slightly outdated but compatible version ?

**Comment by webknjaz** [comment]
@micheelengronne it's not Molecule, it's pip which is managing versions and it will resolve deps accordingly. In case of Ansible, we have `>= 2.5` set. Ansible has its own dist metadata.
Also, I don't understand why would you bring in `requirements.txt` here. That one is for pinning the env deps, not library deps.

**Comment by micheelengronne** [comment]
because I installed all my pip packages with a requirements.txt file. It is more readable and easier to integrate in multiple systems (CI, doc, other projects...).

Does it do the same for ansible-lint ? Does molecule install it ?

I do not have time to check all possible breaking changes for these deps but if I can have my ansible test environment up and running I would be happy.

**Comment by webknjaz** [comment]
> because I installed all my pip packages with a requirements.txt file. It is more readable and easier to integrate in multiple systems (CI, doc, other projects...).

Just to reiterate: libs are distributed as dists, when you install that into a virtualenv you want to lock that env pinning all of the transitive dependencies. That's fine, but use-cases are different. In the dist metadata of Molecule we declare "hey, Molecule needs these things to run properly". Yet, when building a venv for yourself, as a user you want it to be reproducible, but that's out of scope of a single app.

You should use `requirements.txt`, but that's not about the specific lib. It's about your env.

> Does it do the same for ansible-lint ? Does molecule install it ?

Pip installs things, not Molecule. And there's ansible-lint specified as a dependency so it'll be installed. Which doesn't mean that you shouldn't use `requirements.txt`.

You might want to read this first: https://caremad.io/posts/2013/07/setup-vs-requirement/

**Comment by micheelengronne** [comment]
Yes, my bad. I just twisted words. Of course, I know that pip is in charge of installing and that a dist should not pin.

Hey, isn't it the problem here, btw, molecule dist pining too much :)

So ok. I will go with 2 env. One for molecule to test and one for ansible-bender to create Docker images.

**Comment by decentral1se** [comment]
OK, #2025 is merged now. The latest master build should go green over at https://travis-ci.com/ansible/molecule/builds/110410528. You might want to rebase this and let's get this moving forward. Thanks for the patience.

**Comment by decentral1se** [comment]
What's the status here? It should be rebased to remove the duplicated commit of https://github.com/ansible/molecule/pull/2025? I've restarted the CI build in any case ...

File: setup.cfg [hunk 1]
```diff
@@ -71,19 +71,19 @@ install_requires =
     ansible-lint >= 4.0.2, < 5
 
     anyconfig == 0.9.7
-    cerberus >= 1.2, < 3
+    cerberus <= 1.3
```
Comment (webknjaz) [inline setup.cfg]: Please don't remove the lower limit. It's bad.

File: test/unit/model/v2/test_driver_section.py [hunk 2]
```diff
@@ -134,13 +134,7 @@ def _model_driver_provider_name_not_nullable_when_vagrant_section_data():
...
     indirect=True)
 def test_driver_provider_name_not_nullable_when_vagrant_driver(_config):
-    x = {
-        'driver': [{
-            'provider': [{
-                'name': ['unallowed value None', 'null value not allowed']
-            }]
-        }]
-    }
+    x = {'driver': [{'provider': [{'name': ['null value not allowed']}]}]}
```
Comment (webknjaz) [inline test/unit/model/v2/test_driver_section.py]: This change is unrelated.

Comment (micheelengronne) [reply test/unit/model/v2/test_driver_section.py]: I know. It was another issue that blocked the CI.

**Comment by micheelengronne** [comment]
The CI is still blocked and as @webknjaz stated the SemVer of the dependencies is uncertain.

As I didn't check any of them specifically I think it is premature to merge this patch.

File: test/unit/model/v2/test_driver_section.py (hunk 2 above)
Comment (ssbarnea) [reply test/unit/model/v2/test_driver_section.py]: Please avoid touching more than one subject in a PR or it is likely that it will never be merged. Just make another PR for the other fix.

Due to the lack of time I am not even doing reviews that cover move than one area as I know we will not merge them.

Comment (micheelengronne) [reply test/unit/model/v2/test_driver_section.py]: It is not related to this PR at all. I investigated this PR https://github.com/ansible/molecule/issues/2022 in response to this comment https://github.com/ansible/molecule/pull/2024#issuecomment-488607398 and created this MR https://github.com/ansible/molecule/pull/2025 for that. But, as I already said here https://github.com/ansible/molecule/pull/2024#issuecomment-488635238 I am well aware this is not related.

**Comment by ssbarnea** [comment]
Closing as it seems outdated. I think that all cores would agree on "Less constrained dependencies" but this should be done on each dependency based on facts. It makes no sense to add ceiling for packages that never broke anything, like six.
